### PR TITLE
Add chapter 23.

### PR DESCRIPTION
**Pull request by OrangeToucan**

### CODE REVIEW

File: docs/23-mdl-as-a-system-process.markdown [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+This chapter treats MDL considered as executing in an 
+operating-system process, and interactions between MDL and other 
+operating-system processes. See also section 21.8.13.
+
+## 23.1. TIME
+
+`TIME` takes any number of arguments, which are evaluated but 
+ignored, and returns a `FLOAT` giving the number of seconds of CPU 
+time MDL has used so far. `TIMES` is often used in machine-level
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: MDL -> the MDL process
`TIMES` -> `TIME`

File: docs/23-mdl-as-a-system-process.markdown [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+returns a `STRING` which is the "intended user name" of MDL's 
+process. This is the "xuname" process-control variable in the ITS 
+version and identical to <UNAME> in the Tenex and Tops-20 versions.
+
+    <JNAME>
+
+returns a `STRING` which is the "job name" of MDL's process. This is 
+the "jname" process-control variable in the ITS version and the 
+`SETNM` name in the Tenex and Tops-20 versions. The characters belong 
+to the "sixbit" or "printing" subset of ASCI, namely those between
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: ASCI -> ASCII

File: docs/23-mdl-as-a-system-process.markdown [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+returns a `STRING` which is the "job name" of MDL's process. This is 
+the "jname" process-control variable in the ITS version and the 
+`SETNM` name in the Tenex and Tops-20 versions. The characters belong 
+to the "sixbit" or "printing" subset of ASCI, namely those between 
+`<ASCII *40*>` and `<ASCII *137*>` inclusive.
+
+    <XJNAME>
+
+returns a STRING which is the "intended job name" of MDL's process.
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `STRING`

File: docs/23-mdl-as-a-system-process.markdown [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+the "jname" process-control variable in the ITS version and the 
+`SETNM` name in the Tenex and Tops-20 versions. The characters belong 
+to the "sixbit" or "printing" subset of ASCI, namely those between 
+`<ASCII *40*>` and `<ASCII *137*>` inclusive.
+
+    <XJNAME>
+
+returns a STRING which is the "intended job name" of MDL's process. 
+This is the "xjname" process-control variable in the ITS version and 
+identical to <JNAME> in the Tenex and Tops-20 versions.
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `<JNAME>`

File: docs/23-mdl-as-a-system-process.markdown [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+attempts to log out the process in which it is executed. It will 
+succeed only if the MDL is the top-level process, that is, it is 
+running disowned or as a daemon. If it succeeds, it of course never 
+returns. If it does not, it returns `#FALSE ()`.
+
+    <QUIT>
+
+causes MDL to stop running, in an orderly manner. In the ITS version, 
+it is equivalent to a `.LOGOUT` instruction. In the Tenex and Tops-20
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: `.LOGOUT` -> `.LOGOUT 1`

File: docs/23-mdl-as-a-system-process.markdown [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+returns. If it does not, it returns `#FALSE ()`.
+
+    <QUIT>
+
+causes MDL to stop running, in an orderly manner. In the ITS version, 
+it is equivalent to a `.LOGOUT` instruction. In the Tenex and Tops-20 
+versions, it is equivalent to a control-C signal, and control passes 
+to the superior process.
+
+    <VALRET string>
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: string -> string-or-fix

File: docs/23-mdl-as-a-system-process.markdown [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+("value return") seldom returns. It passes control back up the 
+process tree to the superior of MDL, passing its argument as a 
+message to that superior. If it does return, the value is `#FALSE 
+()`. If the argument is a `STRING`, it is passed to the superior as a 
+command to be executed, via `.VALUE` in the ITS version and `RSCAN` 
+in the Tops-20 version. If the argument is a `FIX`, it is passed to 
+the superior as the "effective address" of a .BREAK 16, instruction
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `.BREAK 16,`

File: docs/23-mdl-as-a-system-process.markdown [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+("value return") seldom returns. It passes control back up the 
+process tree to the superior of MDL, passing its argument as a 
+message to that superior. If it does return, the value is `#FALSE 
+()`. If the argument is a `STRING`, it is passed to the superior as a 
+command to be executed, via `.VALUE` in the ITS version and `RSCAN` 
+in the Tops-20 version. If the argument is a `FIX`, it is passed to 
+the superior as the "effective address" of a .BREAK 16, instruction 
+in the ITS version and ignored in other versions.
+
+## 23.4. Interprocess Communication
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: hyphenate "Inter-process"

File: docs/23-mdl-as-a-system-process.markdown [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+All of the SUBRs in this section are available only in the ITS
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `SUBR`

File: docs/23-mdl-as-a-system-process.markdown [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+in the ITS version and ignored in other versions.
+
+## 23.4. Interprocess Communication
+
+All of the SUBRs in this section are available only in the ITS 
+version.
+
+The IPC ("inter-process communication") device is treated as an I/O 
+device by ITS but not explicitly so by MDL: that is, it is never 
+`OPEN`ed. It allowed MDL to communicate with other ITS processes by
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: allowed -> allows

File: docs/23-mdl-as-a-system-process.markdown [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+### 23.4.1. SEND and SEND-WAIT
+
+    <SEND othern1 othern2 body type mynamel myname2>
+
+    <SEND-WAIT othern1 othern2 body type mynamel myname2>
+
+both send an IPC message to any job that is listening for it as 
+*othern1* *othern2*. *body* must be either a `STRING`, or a `UVECTOR` 
+of objects of `PRIMTYPE` `WORD`. *type* is an optional `FIX`, 0 by
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `0`

File: docs/23-mdl-as-a-system-process.markdown [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+(chapter 21.) A handler is called with either four or six arguments 
+cleaned from the message. *body*, *type*, *othern1*, and *othern2* 
+are supplied only if they are not part of this process's `<UNAME>` 
+and `<JNAME>`.
+
+There is a built-in `HANDLER` for the `"IPC"` interrupt, with a 
+handler named `IPC-HANDLER` and `0` in the `PROCESS` slot. The 
+handler prints out on the terminal the *body*, whom it is from, the 
+*type* if not `0`, and whom it is to if not `<UNAME>` `<JNAME>`. If 
+the *type* is 1 and the *body* is a `STRING`, then after the message
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `1`
comma after "then"

File: docs/23-mdl-as-a-system-process.markdown [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+default, which is part of the information the other guy receives. The 
+last two arguments are from whom the message is to be sent. These are 
+optional, and `<UNAME>` and `<JNAME>` respectively are used by 
+default. `SEND` returns a `FALSE` if no one is listening, while 
+`SEND-WAIT` hangs until someone wants it. Both return `T` if someone 
+accepts the message.
+
+### 23.4.2. The "IPC" Interrupt
+
+WHen your MDL process receives an IPC message, `"IPC"` occurs
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: WHen -> When

File: docs/23-mdl-as-a-system-process.markdown [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+last two arguments are from whom the message is to be sent. These are 
+optional, and `<UNAME>` and `<JNAME>` respectively are used by 
+default. `SEND` returns a `FALSE` if no one is listening, while 
+`SEND-WAIT` hangs until someone wants it. Both return `T` if someone 
+accepts the message.
+
+### 23.4.2. The "IPC" Interrupt
+
+WHen your MDL process receives an IPC message, `"IPC"` occurs 
+(chapter 21.) A handler is called with either four or six arguments
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: period outside parens

File: docs/23-mdl-as-a-system-process.markdown [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+optional, and `<UNAME>` and `<JNAME>` respectively are used by 
+default. `SEND` returns a `FALSE` if no one is listening, while 
+`SEND-WAIT` hangs until someone wants it. Both return `T` if someone 
+accepts the message.
+
+### 23.4.2. The "IPC" Interrupt
+
+WHen your MDL process receives an IPC message, `"IPC"` occurs 
+(chapter 21.) A handler is called with either four or six arguments 
+cleaned from the message. *body*, *type*, *othern1*, and *othern2*
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: cleaned -> gleaned

File: docs/23-mdl-as-a-system-process.markdown [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+default. `SEND` returns a `FALSE` if no one is listening, while 
+`SEND-WAIT` hangs until someone wants it. Both return `T` if someone 
+accepts the message.
+
+### 23.4.2. The "IPC" Interrupt
+
+WHen your MDL process receives an IPC message, `"IPC"` occurs 
+(chapter 21.) A handler is called with either four or six arguments 
+cleaned from the message. *body*, *type*, *othern1*, and *othern2* 
+are supplied only if they are not part of this process's `<UNAME>`
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: delete "part of"

File: docs/23-mdl-as-a-system-process.markdown [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+There is a built-in `HANDLER` for the `"IPC"` interrupt, with a 
+handler named `IPC-HANDLER` and `0` in the `PROCESS` slot. The 
+handler prints out on the terminal the *body*, whom it is from, the 
+*type* if not `0`, and whom it is to if not `<UNAME>` `<JNAME>`. If 
+the *type* is 1 and the *body* is a `STRING`, then after the message 
+information is printed out, the `STRING` is `PARSE`d and `EVAL`uated.
+
+### 23.4.3. IPC-OFF
+
+`<IPC-OFF>` stops listening on the IPC device.
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: stops listening -> stops all listening

File: docs/23-mdl-as-a-system-process.markdown [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+### 23.4.4. IPC-ON
+
+    <IPC-ON myname1 myname2>
+
+causes listening on the IPC device as *myname1* *myname2*. If no 
+arguments are provided, listening is on `<UNAME>` `<JNAME>`. When a 
+message arrives, `"IPC"` occurs.
+
+MDL is initially listening as `<UNAME>` `<JNAME>` with the built-in 
+`HANDLER` set up on the `"IPC"` interrupt with a priority of 1.
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: backtick `1`?

**Comment by OrangeToucan** [comment]
I have made these changes.

File: docs/23-mdl-as-a-system-process.markdown [hunk 19]
```diff
@@ -0,0 +1,137 @@
+### Chapter 23. MDL as a System Process
```
Comment (taradinoc) [inline docs/23-mdl-as-a-system-process.markdown]: level 1 heading

**Comment by OrangeToucan** [comment]
Also updating mkdocs.yml may be reaching the limits of my git knowledge, because at the time that the branch was created chapter 22 was not in there yet which meant I couldn't add chapter 23 and now I don't know how to do that.

**Comment by OrangeToucan** [comment]
Supposedly I tried pulling in your addition of chapter 22 into my fork with a pull request so that I could rebase my chapter-23 branch on the updated master but now these changes seem to be mentioned in this pull request, and my chapter-23 branch still does not seem updated, even though I did did rebase master, made changes, and did git commit --amend followed by git push --force origin chapter-23.

I suppose I should have only tried to merge the chapters one at a time. Now everything seems to be falling all over itself.

**Comment by OrangeToucan** [comment]
After more git head banging I'm going to close these pull requests, delete everything, and start over. Very much a case of https://xkcd.com/1597/ (mkdocs.yml could have been updated later.)

**Pull request closed without merging**